stdlib/native.rb

```diff
@@ -160,6 +160,44 @@ def Array(object, *args, &block)
       }
     }
   end
+  
+  # allows to pass a proc that runs in context of `this`
+  #
+  # @example
+  #   options = {
+  #     persist: false,
+  #     render: function { |item, escape|
+  #       '<div>' + escape.call(item.name) + '</div>'
+  #     }
+  #   }
+  #
+  def function *args, &block
+    args.any? && raise(ArgumentError, '`function` does not accept arguments')
+    block || raise(ArgumentError, 'block required')
+    proc do |*a| # no way to expect a block here as this is always called from JavaScript
+      a.map! {|x| Native(`x`)}
+      instance = Native(`this`)
+      %x{
+        // if window is current scope, run the block in the scope it was defined
+        if(this.toString() === '[object Window]') {
+          return block.apply(self, a);
```

Comment:
Right, but `self` is a keyword in Ruby, not a local variable in JavaScript (unless that's how Opal implements it?). So since this is in an x-string and you're referencing JavaScript `self`, not Ruby `self`...

Sorry for all the question. Again, I'm not familiar with Opal's internals.